documentation/doxygen/man/man3/Client.3
.TH "Client" 3 "Tue Jan 9 2024" "kOrganizify" \" -*- nroff -*-
.ad l
.nh
.SH NAME
Client
.SH SYNOPSIS
.br
.PP
.PP
\fC#include <client\&.h>\fP
.PP
Inherits QObject\&.
.SS "Public Slots"

.in +1c
.ti -1c
.RI "void \fBreadFromServer\fP ()"
.br
.RI "\fBClient\fP can receive one of the following messages: "
.ti -1c
.RI "void \fBdisconnected\fP ()"
.br
.ti -1c
.RI "void \fBsyncResponse\fP (bool response, QString username, QString friendName, int duration, \fBCalendar\fP cal={}) const"
.br
.ti -1c
.RI "void \fBeventResponse\fP (bool response)"
.br
.ti -1c
.RI "void \fBsyncRequest\fP (QString from, QString to, QString titleEvent, int duration, \fBCalendar\fP calendar)"
.br
.in -1c
.SS "Signals"

.in +1c
.ti -1c
.RI "void \fBnewUserLoggedIn\fP (const QString &username)"
.br
.ti -1c
.RI "void \fBdisconnectedUser\fP (const QString &username)"
.br
.ti -1c
.RI "void \fBshowSyncWindow\fP (QString from, QString eventTitle, int duration)"
.br
.ti -1c
.RI "void \fBsyncRequestDenied\fP (QString friendName)"
.br
.ti -1c
.RI "void \fBsyncSuccess\fP (QDateTime startTime, QDateTime endTime, QString eventTitle)"
.br
.ti -1c
.RI "void \fBnewEventSync\fP (QString eventTitle, QString startTime)"
.br
.in -1c
.SS "Public Member Functions"

.in +1c
.ti -1c
.RI "\fBClient\fP (QString username, QObject *parent=nullptr)"
.br
.ti -1c
.RI "void \fBlogoutHappened\fP (QString username)"
.br
.in -1c
.SS "Public Attributes"

.in +1c
.ti -1c
.RI "QString \fBm_username\fP"
.br
.ti -1c
.RI "QList< QString > \fBm_friends\fP"
.br
.in -1c
.SS "Private Member Functions"

.in +1c
.ti -1c
.RI "void \fBsendMessage\fP (QString message)"
.br
.ti -1c
.RI "bool \fBmakeConnection\fP (QHostAddress::SpecialAddress address)"
.br
.in -1c
.SS "Private Attributes"

.in +1c
.ti -1c
.RI "QTcpSocket * \fBm_socket\fP"
.br
.in -1c
.SH "Constructor & Destructor Documentation"
.PP 
.SS "Client::Client (QString username, QObject * parent = \fCnullptr\fP)"

.SH "Member Function Documentation"
.PP 
.SS "void Client::disconnected ()\fC [slot]\fP"

.SS "void Client::disconnectedUser (const QString & username)\fC [signal]\fP"

.SS "void Client::eventResponse (bool response)\fC [slot]\fP"

.SS "void Client::logoutHappened (QString username)"

.SS "bool Client::makeConnection (QHostAddress::SpecialAddress address)\fC [private]\fP"

.SS "void Client::newEventSync (QString eventTitle, QString startTime)\fC [signal]\fP"

.SS "void Client::newUserLoggedIn (const QString & username)\fC [signal]\fP"

.SS "void Client::readFromServer ()\fC [slot]\fP"

.PP
\fBClient\fP can receive one of the following messages: 
.IP "1." 4
new connection
.IP "2." 4
disconnected
.IP "3." 4
syncRequest
.IP "4." 4
rejectSync
.IP "5." 4
new sync event
.IP "6." 4
agreed sync 
.PP

.SS "void Client::sendMessage (QString message)\fC [private]\fP"

.SS "void Client::showSyncWindow (QString from, QString eventTitle, int duration)\fC [signal]\fP"

.SS "void Client::syncRequest (QString from, QString to, QString titleEvent, int duration, \fBCalendar\fP calendar)\fC [slot]\fP"

.SS "void Client::syncRequestDenied (QString friendName)\fC [signal]\fP"

.SS "void Client::syncResponse (bool response, QString username, QString friendName, int duration, \fBCalendar\fP cal = \fC{}\fP) const\fC [slot]\fP"

.SS "void Client::syncSuccess (QDateTime startTime, QDateTime endTime, QString eventTitle)\fC [signal]\fP"

.SH "Member Data Documentation"
.PP 
.SS "QList<QString> Client::m_friends"

.SS "QTcpSocket* Client::m_socket\fC [private]\fP"

.SS "QString Client::m_username"


.SH "Author"
.PP 
Generated automatically by Doxygen for kOrganizify from the source code\&.


kOrganizify/src/client.cpp
#include "client.h"
#include "event.h"
#include <QCoreApplication>
#include <QEventLoop>
#include <QJsonDocument>
#include <QJsonObject>
#include <utility>

Client::Client(QString username, QObject *parent) : QObject(parent), m_username(std::move(username)) {
    m_socket = new QTcpSocket(this);
    makeConnection(QHostAddress::LocalHost);
}

auto Client::makeConnection(QHostAddress::SpecialAddress address) -> bool {
    connect(m_socket, &QTcpSocket::readyRead, this, &Client::readFromServer);
    connect(m_socket, &QTcpSocket::disconnected, this, &Client::disconnected);

    m_socket->connectToHost(QHostAddress(address), 12345);
    if (m_socket->waitForConnected(5000)) {
        qDebug() << "Connected to the server.";
    } else {
        qDebug() << "Connection failed. Error:" << m_socket->error();
        return false;
    }
    QJsonObject newClientMessage;
    newClientMessage.insert("title", "new connection");
    newClientMessage.insert("username", m_username);
    QString msg(QJsonDocument(newClientMessage).toJson());
    m_socket->write(msg.toStdString().c_str());
    return m_socket->waitForConnected();
}

void Client::disconnected() {
    delete m_socket;
    QCoreApplication::quit();
}

/// Client can receive one of the following messages:
/// 1. new connection
/// 2. disconnected
/// 3. syncRequest
/// 4. rejectSync
/// 5. new sync event
/// 6. agreed sync
void Client::readFromServer() {
    QString message = m_socket->readAll();
    QJsonObject doc = QJsonDocument::fromJson(message.toUtf8()).object();
    QString title   = doc.value("title").toString();
    if (title == "new connection") {
        m_friends.append(doc.value("username").toString());
        emit newUserLoggedIn(doc.value("username").toString());
    } else if (title == "disconnected") {
        m_friends.removeOne(doc.value("username").toString());
        emit disconnectedUser(doc.value("username").toString());
    } else if (title == "syncRequest") {
        QString from       = doc.value("fromUsername").toString();
        QString eventTitle = doc.value("titleEvent").toString();
        int duration       = doc.value("duration").toString().toInt();

        emit showSyncWindow(from, eventTitle, duration);
    } else if (title == "rejectSync") {
        QString friendName = doc.value("fromUsername").toString();
        emit syncRequestDenied(friendName);
    } else if (title == "new sync event") {
        Event event;
        event.setTitle(doc.value("eventTitle").toString());
        event.setStartTime(QDateTime::fromString(doc.value("startTime").toString(), Qt::ISODate));

        emit newEventSync(doc.value("eventTitle").toString(), doc.value("startTime").toString());
    } else if (title == "agreed sync") {
        QDateTime startTime = QDateTime::fromString(doc.value("startTime").toString()); // won't convert
        QDateTime endTime   = QDateTime::fromString(doc.value("endTime").toString());
        QString title       = doc.value("eventTitle").toString();

        emit syncSuccess(startTime, endTime, title);
    }
}

void Client::sendMessage(QString x) {
    QByteArray toSend = QByteArray::fromStdString(x.toStdString());

    m_socket->write(toSend);
}

void Client::syncResponse(bool response, QString username, QString friendName, int duration, Calendar cal) const {
    if (!response) {
        QJsonObject newClientMessage;
        newClientMessage.insert("title", "rejectSync");
        newClientMessage.insert("fromUsername", username);
        newClientMessage.insert("toUsername", friendName);
        QString msg(QJsonDocument(newClientMessage).toJson());
        m_socket->write(msg.toStdString().c_str());
    } else {
        QJsonArray jsonArray;
        for (const Event &event : cal.getEvents()) {
            QJsonObject jsonObject;
            jsonObject["title"]       = event.getTitle();
            jsonObject["startTime"]   = event.getStartTime().toString(Qt::ISODate);
            jsonObject["endTime"]     = event.getEndTime().toString(Qt::ISODate);
            jsonObject["description"] = event.getDescription();
            jsonObject["location"]    = event.getLocation();

            jsonArray.append(jsonObject);
        }

        QJsonDocument doc;
        doc.setArray(jsonArray);
        QString dataToString(doc.toJson());

        QJsonObject newClientMessage;
        newClientMessage.insert("title", "acceptSync");
        newClientMessage.insert("fromUsername", username);
        newClientMessage.insert("toUsername", friendName);
        newClientMessage.insert("syncEventDuration", duration);
        newClientMessage.insert("events", QJsonValue(jsonArray));
        QString msg(QJsonDocument(newClientMessage).toJson());
        m_socket->write(msg.toStdString().c_str());
    }
}

void Client::eventResponse(bool response) {
    QJsonObject responseMsg;
    responseMsg.insert("title", "eventResponse");
    responseMsg.insert("answer", response ? "yes" : "no");
    QString msg(QJsonDocument(responseMsg).toJson());

    m_socket->write(msg.toStdString().c_str());
    m_socket->flush();
}

void Client::syncRequest(QString from, QString to, QString titleEvent, int duration, Calendar calendar) {
    QJsonObject responseMsg;
    responseMsg.insert("title", "syncRequest");
    responseMsg.insert("fromUsername", from);
    responseMsg.insert("toUsername", to);
    responseMsg.insert("titleEvent", titleEvent);
    responseMsg.insert("duration", QString::number(duration));

    QJsonArray jsonArray;
    for (const Event &event : calendar.getEvents()) {
        QJsonObject jsonObject;
        jsonObject["title"]       = event.getTitle();
        jsonObject["startTime"]   = event.getStartTime().toString(Qt::ISODate);
        jsonObject["endTime"]     = event.getEndTime().toString(Qt::ISODate);
        jsonObject["description"] = event.getDescription();
        jsonObject["location"]    = event.getLocation();

        jsonArray.append(jsonObject);
    }

    responseMsg.insert("events", QJsonValue(jsonArray));
    QString msg(QJsonDocument(responseMsg).toJson());

    m_socket->write(msg.toStdString().c_str());
}

void Client::logoutHappened(QString username) {
    QJsonObject responseMsg;
    responseMsg.insert("title", "logout");
    responseMsg.insert("username", username);
    QString msg(QJsonDocument(responseMsg).toJson());

    m_socket->write(msg.toStdString().c_str());
    m_socket->flush();
}
